### ISSUE TITLE
Re-running make -j12 hangs using roslyn compile server

### ISSUE THREAD

**Issue by lambdageek** [mono/mono#15267]
This is with mono a4ef5c3590b8a86e4e4827b50bf923108db8f7ac on OSX.

## Steps to Reproduce

1. `./autogen.sh  --enable-checked-build-mode=all --disable-btls --disable-btls-lib --disable-nls`
2. `make -j12`  This one works normally.
3. `mv mcs/build/compile-server.log compile-server.log.good`
4. `ps awx |grep -i VBCS` there should be no compile server running
5. `make -j12` This one hangs
6. Ctrl-C
7. `ps awx | grep -i VBCS` note a running process: `/Users/alklig/work/mono/mono/mini/mono --config /Users/alklig/work/mono/runtime/etc/mono/config --gc-params=nursery-size=64m /Users/alklig/work/mono/external/roslyn-binaries/Microsoft.Net.Compilers/3.1.0-release/VBCSCompiler.exe -pipename:monomake`

### Current Behavior

Mono starts the compile server and then hangs after the build is done.

### Expected Behavior

Mono doesn't hang.

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

building from git checkout of a4ef5c3590b8a86e4e4827b50bf923108db8f7ac

## Stacktrace

No crash.

Attaching good build log 
[aleksey-build.log](https://github.com/mono/mono/files/3307736/aleksey-build.log)

Good compile log 
[compile-server.log.clean.txt](https://github.com/mono/mono/files/3307739/compile-server.log.clean.txt)

Second run build log

[aleksey-build.log.rerun.txt](https://github.com/mono/mono/files/3307742/aleksey-build.log.rerun.txt)

Hung compile log

[compiler-server.log](https://github.com/mono/mono/files/3307743/compiler-server.log)

**Comment by CoffeeFlux** [comment]
I've had this happen rarely and not been able to reliably reproduce it - simply restarting the build process makes it go away. Meant to file an issue last time it happened but got distracted :(

If it happens again, I'll upload logs here as well.

**Comment by kg** [comment]
When this happens do you end up with two instances of VBCSCompiler at the end? Or just one?

**Comment by lambdageek** [comment]
@kg just one.

**Comment by kg** [comment]
I spent a while trying to reproduce this today and wasn't able to - I left make -j 12 going in a loop for about 8 hours. I've seen it once before so it's definitely a bug but I'm not sure how to investigate it unless we have a machine in this bad state. If you're working in a VM could you suspend and take a snapshot the next time this happens so I can try to boot up the VM and dig in?

make hanging on the child process is definitely bad and shouldn't be happening, so that sounds like Unix Nonsense. I know there are a few circumstances under which that can happen but I haven't found any examples of it w/make on SO or discussion lists...

**Comment by CoffeeFlux** [comment]
Just ran into this again myself. I can't offer reliable repro steps, though in my case I got it from making a netcore build. Not much interesting in the build log, but I've uploaded it regardless: https://gist.github.com/CoffeeFlux/70830fff8f4322149908ba6ff43523b0

Running `ps aux | grep VBCS` gave me `ryan             90924   0.0  0.4 23391932 120976 s001  S+    5:02PM   0:33.27 /Users/ryan/git/mono/.dotnet/dotnet exec /Users/ryan/git/mono/netcore/roslyn/packages/microsoft.net.compilers.toolset/3.2.0-beta3-19310-04/tasks/netcoreapp2.1/bincore/VBCSCompiler.dll -pipename:RBWD31Z+njQv7u8pdvLqixyb9f73VYF0Aurx2SxG0gk`, and then after that timed out in I believe 10 minutes, at which point make finished up.

**Comment by kg** [comment]
> 
> Just ran into this again myself. I can't offer reliable repro steps, though in my case I got it from making a netcore build. Not much interesting in the build log, but I've uploaded it regardless: https://gist.github.com/CoffeeFlux/70830fff8f4322149908ba6ff43523b0
> 
> Running `ps aux | grep VBCS` gave me `ryan 90924 0.0 0.4 23391932 120976 s001 S+ 5:02PM 0:33.27 /Users/ryan/git/mono/.dotnet/dotnet exec /Users/ryan/git/mono/netcore/roslyn/packages/microsoft.net.compilers.toolset/3.2.0-beta3-19310-04/tasks/netcoreapp2.1/bincore/VBCSCompiler.dll -pipename:RBWD31Z+njQv7u8pdvLqixyb9f73VYF0Aurx2SxG0gk`, and then after that timed out in I believe 10 minutes, at which point make finished up.
> 
> Edit: Should note I'm on osx as well

OK, so by 'timed out' you mean VBCS automatically exited?

**Comment by CoffeeFlux** [comment]
Yeah, automatically exited. The log is long gone; I thought I'd included it in my comment but apparently not, sorry. It would be nice if I could get this to happen in a VM so I could just give you an image, but so far it's only happened sporadically on OSX...
